www/js/app.js

```diff
@@ -569,19 +651,45 @@ define(['jquery', 'zimArchiveLoader', 'uiUtil', 'settingsStore','abstractFilesys
      * @param {String} value The chosen content injection mode : 'jquery' or 'serviceworker'
      */
     function setContentInjectionMode(value) {
+        params.contentInjectionMode = value;
         if (value === 'jquery') {
-            if (isServiceWorkerReady()) {
-                // We need to disable the ServiceWorker
-                // Unregistering it does not seem to work as expected : the ServiceWorker
-                // is indeed unregistered but still active...
-                // So we have to disable it manually (even if it's still registered and active)
-                navigator.serviceWorker.controller.postMessage({'action': 'disable'});
-                messageChannel = null;
+            if (params.extensionURL) {
+                // We are in an extension, and the user may wish to revert to local code
+                var message = 'This will switch to using locally packaged code only. Some configuration settings may be lost.\n\n' +
+                'WARNING: After this, you may not be able to switch back to SW mode without an online connection!';
+                var launchLocal = function () {
+                    settingsStore.setItem('allowInternetAccess', false, Infinity);
+                    var uriParams = '?allowInternetAccess=false&contentInjectionMode=jquery&hideActiveContentWarning=false';
+                    uriParams += '&appTheme=' + params.appTheme;
+                    uriParams += '&showUIAnimations=' + params.showUIAnimations; 
+                    window.location.href = params.extensionURL + '/www/index.html' + uriParams;
+                    'Beam me down, Scotty!';
+                };
+                var response = confirm(message);
+                if (response) {
+                    launchLocal();
+                } else {
+                    setContentInjectionMode('serviceworker');
+                }
+                return;
+            }
+            // Because the "outer" Service Worker still runs in a PWA app, we don't actually disable the SW in this context, but it will no longer
+            // be intercepting requests
+            if ('serviceWorker' in navigator) {
+                serviceWorkerRegistration = null;
             }
             refreshAPIStatus();
-            // User has switched to jQuery mode, so no longer needs CACHE_NAME
-            // We should empty it to prevent unnecessary space usage
-            if ('caches' in window) caches.delete(CACHE_NAME);
+            // User has switched to jQuery mode, so no longer needs ASSETS_CACHE
+            // We should empty it and turn it off to prevent unnecessary space usage
+            if ('caches' in window && isMessageChannelAvailable()) {
+                var channel = new MessageChannel();
+                if (isServiceWorkerAvailable && navigator.serviceWorker.controller) {
```

Comment:
This line generates a blocking error on browsers that do not support ServiceWorkers (like my Firefox OS device).
It can be easily fixed this way : 
```suggestion
                if (isServiceWorkerAvailable && navigator.serviceWorker && navigator.serviceWorker.controller) {
```

Reply:
Apart from that, I did not see any regression on my Firefox OS device.
Of course, as it does not support Serviceworkers, I could not test this feature on it at all

Reply:
OK, that's helpful, thank you.

Reply:
It seems anomalous that `isServiceWorkerAvailable` returns true on this device, but `navigator.serviceWorker` returns false!

Reply:
That's right!
In fact, it's because the existing code was not calling the function `isServiceWorkerAvailable`, but only testing it existed...

So I suggest this fix instead (tested on my Firefox OS device)

```suggestion
                if (isServiceWorkerAvailable() && navigator.serviceWorker.controller) {
```

Reply:
Oops, silly me! Thanks for spotting that.